frameworklauncher/src/main/java/com/microsoft/frameworklauncher/webserver/LauncherModule.java

```diff
@@ -65,15 +68,89 @@ private static LaunchClientType getLaunchClientType(
     }
   }
 
-  private static String getUserName(
-      CommonExts.NoExceptionCallable<String> ResolveUserName) throws BadRequestException {
-    String userName = ResolveUserName.call();
-    if (userName == null) {
+  private static String getName(
+      CommonExts.NoExceptionCallable<String> ResolveName) throws BadRequestException {
+    String name = ResolveName.call();
+    if (name == null) {
       return null;
     }
 
-    CommonValidation.validate(userName);
-    return userName;
+    CommonValidation.validate(name);
+    return name;
+  }
+
+  private static Boolean getBoolean(
+      CommonExts.NoExceptionCallable<String> ResolveBoolean) throws BadRequestException {
+    String booleanStr = ResolveBoolean.call();
+    if (booleanStr == null) {
+      return null;
+    }
+
+    if (!booleanStr.equalsIgnoreCase("true") && !booleanStr.equalsIgnoreCase("false")) {
+      throw new BadRequestException(String.format(
+          "Failed to ParseBooleanStr: [%s]", booleanStr));
+    }
+
+    return Boolean.valueOf(booleanStr);
+  }
+
+  private void checkWritableAccess(
+      HttpServletRequest hsr) throws Exception {
+    checkWritableAccess(hsr, null, null);
+  }
+
+  private void checkWritableAccess(
+      HttpServletRequest hsr,
+      String frameworkName) throws Exception {
+    checkWritableAccess(hsr, frameworkName, null);
+  }
+
+  private void checkWritableAccess(
+      HttpServletRequest hsr,
+      String frameworkName,
+      UserDescriptor frameworkUser) throws Exception {
+    if (!conf.getWebServerAclEnable()) {
+      return;
+    }
+
+    // Check Admin
+    Boolean isAdmin = getBoolean(() -> hsr.getHeader(WebCommon.REQUEST_HEADER_IS_ADMIN));
+    if (isAdmin != null && isAdmin) {
+      // Admin Users can always access anything
+      return;
+    } else if (frameworkName == null) {
+      // Non Admin can only write Framework Request
+      throw new AuthorizationException(
+          "This operation needs administrator privilege.");
+    }
+
+    UserDescriptor user = UserDescriptor.newInstance(
+        getName(() -> hsr.getHeader(WebCommon.REQUEST_HEADER_USER_NAME)));
+
+    // Check UserName consistency
+    if (frameworkUser != null && !frameworkUser.equals(user)) {
+      throw new AuthorizationException(String.format(
+          "User [%s] cannot submit a Framework which belongs to another User [%s].",
+          user, frameworkUser));
+    }
+
+    // Check Namespace Acl
+    String namespace = CommonValidation.validateAndGetNamespace(frameworkName);
+    AccessControlList namespaceAcl = new AccessControlList();
+
+    // 1. Add Predefined Acl
+    // 1.1 User can always write the Namespace whose name is the same as User
+    namespaceAcl.addUser(UserDescriptor.newInstance(namespace));
+    // 2. Add Configured Acl
+    namespaceAcl.addUsers(requestManager.getLauncherRequest().
+        getAclConfiguration().getNamespaceAcl(namespace).getUsers());
+
```

Comment:
Ever considered reusing the original acl object for verification instead of generating temporary objects? This function might called frequently.

Reply:
No, it will changes the persistent info: users for one queue.
For example, if we add one user here, it actually modified the underlay persistent info from request manager. It will bring the whole program to an undefined behavior.
BTW, checkWritableAccess is not a bottleneck for Writable RestAPI call here, ZK access is the bottleneck.

---
In reply to: [167464437](https://github.com/Microsoft/pai/pull/320#discussion_r167464437) [](ancestors = 167464437)

Reply:
So suggest encapsulate the operation into a common function.

---
In reply to: [167466563](https://github.com/Microsoft/pai/pull/320#discussion_r167466563) [](ancestors = 167466563,167464437)